### DISCUSSION TITLE
ID / Name label option for Native Filters

### DISCUSSION THREAD

**Discussion by polmonso** [apache/superset#26084] [Ideas]
**Is your feature request related to a problem? Please describe.**

In many DBs there's an ID / Name pattern. So the <em>id</em> is integer and <em>name</em> is String for human readable representation.

When using Native Filters you have to decide to use id field or String field,
 but ideally you'd like to use the id field for the technical filter and
 use the name for labelling only.

example table:

id | name | revenue
-- | -- | --
1 | sports | 1000
2 | cars | 3000
3 | entertainment | 20000

<p dir="auto">Benefits would be:</p><ul dir="auto">
<li>User still sees a list of Strings to include (as if he selected <em>name</em>)</li>
<li><em>preselecxt_fiter</em> option could inject INTEGER values from 
the outside. This is especially useful from an integration perspective 
as normally APIs and applications that link directly into superset and 
want to preselect a filter will work on the ID rather than on the name</li>
<li>imagine the name is changed -this would break the 'contract' between
 the url parameters and the DB values. The argument, it's based on the 
data is very weak as names are considered dimension data and thus any 
data tool should rely on the ID field.</li>

</ul><p dir="auto"><strong>Describe the solution you'd like</strong><br>
It would be great to have an option 'verbose column' (or similar) in the
 native filter's parameters. That column selection is optional, but would 
take care about the  ID -&gt; string mapping.</p><p dir="auto"><a target="_blank" rel="noopener noreferrer nofollow" href="https://user-images.githubusercontent.com/4490840/109569700-1ef0d380-7ae9-11eb-9d3c-8f5135225a87.png"><img width="390" alt="Screenshot 2021-03-01 at 23 51 10" src="https://user-images.githubusercontent.com/4490840/109569700-1ef0d380-7ae9-11eb-9d3c-8f5135225a87.png" style="max-width: 100%;"></a></p><p dir="auto">In
 the UI filter box a user would just see values like 'sports', 'cars', 
'entertainment', but it could be controlled by preselect_filter using 
IDs 1, 2, 3.<br>
<a target="_blank" rel="noopener noreferrer nofollow" href="https://user-images.githubusercontent.com/4490840/109570645-a9860280-7aea-11eb-82ae-d0111c117858.png"><img width="326" alt="Screenshot 2021-03-02 at 00 02 26" src="https://user-images.githubusercontent.com/4490840/109570645-a9860280-7aea-11eb-82ae-d0111c117858.png" style="max-width: 100%;"></a></p><p dir="auto"><strong>Describe alternatives you've considered</strong><br>

If the 'verbose' field could be entered using a dedicated SQL expression
 potentially controlled/secured by some JINJA templating, that would 
maby more technical, but also more flexible.

This discussion is a copy of https://github.com/apache/superset/discussions/19184 but updated to Native Filters. I didn't update the images, if it helps spring the discussion I'd do some mock ups.

**Comment by jaksmid** [comment]
Great idea, would make life a lot easier

**Comment by dimoha** [comment]
An additional and very important advantage of this approach is that, as it seems to me, there are usually indexes on IDs, and filtering by string will usually occur without an index

**Comment by linhmiyano99** [comment]
sound great, will you implement this feature soon ? @polmonso

**Comment by polmonso** [comment]
@linhmiyano99 I don't think I'll have time at the moment... Anybody can take it.

**Comment by TechAuditBI** [comment]
Well we've already implemented what looks like this feature almost exactly. I think we could contribute it to the upstream.

**Reply by ttauzo** [reply]
In which version it is available?

**Reply by TechAuditBI** [reply]
As for right now it is not available yet. I'll post a link to a PR as it is created. But you probably can expect this feature to appear in 4.0.1 and maybe in 3.1.2 but you should probably ask community member for a more certain answer.

**Reply by micode-ai** [reply]
> As for right now it is not available yet. I'll post a link to a PR as it is created. But you probably can expect this feature to appear in 4.0.1 and maybe in 3.1.2 but you should probably ask community member for a more certain answer.

Has this update become available in any of the versions?

**Reply by LasseLegarth** [reply]
I can't find this in 4.0.1 - Any roadmap for when it will be here?

**Reply by leschiffres** [reply]
Any idea when this feature will be available?

**Reply by icrc-fdeniger** [reply]
Hello, is there any news on this feature ?

**Comment by nmella** [comment]
Can this be achieved using JINJA templating ?

**Comment by Aleksei-Kuz** [comment]
Agree, that's feature is needed.

**Comment by gobeah** [comment]
The feature is not implemented yet, but you can use a hack.
You have 'channels' table:
| id | name |
|--------|--------|
| 1 | sports |
| 2 | cars |
| 3 | entertainment | 

And you have 'news' table:
| id | title | channel_id |
|--------|--------|--------|
| 1 | My first car | 2 |
| 2 | He scores! | 1 |
| 3 | World cup | 1 |
| 4 | Car sale | 2 |
| 5 | Championship 2025 | 1 |
| 6 | Pool party | 3 | 

### Solution
1. Go to SQL Lab and enter this query:
`SELECT`
`  id,`
`  title,`
`  ch.name AS channel_name`
`FROM news n`
`LEFT JOIN channels ch ON n.channel_id = ch.id`
2. Save query as new dataset 'news_extended'.
3. Create new chart (e.g. table) using 'news_extended' dataset.
4. Place chart on dashboard.
5. Add new native filter with settings:
- filter type: Value
- filter name: Channels
- dataset: news_extended
- column: channel_name
6. Save filter and enjoy.
![photo_2025-06-06 22 33 17](https://github.com/user-attachments/assets/d98748c8-cd22-46d8-9ddd-c4be52be75d0)

But I'm afraid it will work slowly, because first all the data will be fetched, and then filter by channel_name will be performed.

p.s. it is important to use 'news_extended' dataset for all charts and filters to make it work.

**Comment by AurVal** [comment]
Very useful feature. Is there any news/plans about this?

**Comment by durlabhjain** [comment]
IT is surprising that such a basic feature is missing. We have a lot of filters and the datasets should ideally be filtered by the ID field while user sees the Verbose (or Display) field to make it more user friendly.

**Comment by durlabhjain** [comment]
I've been able to get my local instance on 6.1.0 to have this option. https://github.com/Coolr-Group-Inc/superset-plugin-filter-value-label

After creating this plugin - based on the native SelectFilter, we had to make a few changes:

1. useFilterOperations.ts - add filter_value_label in ALLOW_DEPENDENCIES
2. superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/getControlItemsMap.tsx - Add code to allow custom filters (https://github.com/Coolr-Group-Inc/superset-plugin-filter-value-label/blob/master/superset-isColumnSelect.patch)

Then we just registered the plugin and it works well. I'm not sure if development team will consider making filter plugins extensible by adding:

- Ability to register additional filters to ALLOW_DEPENDENCIES when a filter plugin is added
- Ability to have custom form items similar to the one we added in the plugin

Any feedback is welcome. I hope it helps others as well.

**Comment by durlabhjain** [comment]
There is a patch https://github.com/apache/superset/pull/40905 that makes adding such plugins easier - without having to patch internal files. Hopefully this gets merged soon
